test: some more tests for django

gauravsnj opened the pull request.

gauravsnj committed:
added some tests

gauravsnj committed:
added tests for selecting all null and error in transaction

gauravsnj committed:
Merge branch 'postgresql-dialect' into django-extra-tests

gauravsnj committed:
formatted the files

gauravsnj committed:
Merge branch 'postgresql-dialect' into django-extra-tests

olavloite commented on src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoBasicTest.java:
```diff
@@ -230,4 +265,24 @@ public void testInsertAllTypesAllNull() throws IOException, InterruptedException
   }
+
+  @Test
+  public void testSelectAllNull() throws IOException, InterruptedException {
```
We should test that null values are correctly mapped for all possible data types, not only for strings. So we should create an ORM model / class that has all data types and then verify that we can select an instance of that model when:
1. All values are non-null
2. All values are null

See for example this test for `gorm`:
1. `gorm` struct for an entity with all data types: https://github.com/GoogleCloudPlatform/pgadapter/blob/8a19695373ed5371e9baea4a5980b10c50736403/src/test/golang/pgadapter_gorm_tests/gorm.go#L65
2. A test that verifies that selecting an instance of that entity with all null values works as expected: https://github.com/GoogleCloudPlatform/pgadapter/blob/8a19695373ed5371e9baea4a5980b10c50736403/src/test/java/com/google/cloud/spanner/pgadapter/golang/GormMockServerTest.java#L333

gauravsnj replied:
done

olavloite commented on src/test/python/django/transaction_test.py:
```diff
@@ -80,6 +80,20 @@ def test_nested_atomic():
+
+    singer.save()
+
+    raise Exception("Some error has occurred")
```
We should not raise an exception in the application itself, but simulate an error in Cloud Spanner. The reason for this is that PostgreSQL has a very strict model when it comes to handling database errors in transactions:
1. A database error in a transaction will invalidate that transaction, and all further statements in that transaction will also fail with the error `current transaction is aborted, commands ignored until end of transaction block`.
2. The only way to get out of that state is to roll back the transaction (or roll back to a savepoint in the transaction, but that is not supported in PGAdapter at the moment).

We should therefore verify that PGAdapter in combination with DJango really shows that behavior.

See this example:
1. `gorm` code: https://github.com/GoogleCloudPlatform/pgadapter/blob/8a19695373ed5371e9baea4a5980b10c50736403/src/test/golang/pgadapter_gorm_tests/gorm.go#L527
2. Corresponding Java test case: https://github.com/GoogleCloudPlatform/pgadapter/blob/8a19695373ed5371e9baea4a5980b10c50736403/src/test/java/com/google/cloud/spanner/pgadapter/golang/GormMockServerTest.java#L780

gauravsnj replied:
Okay got it. Changed the code.

gauravsnj committed:
changed transaction error test

gauravsnj committed:
Update DjangoTransactionsTest.java

gauravsnj committed:
updated code to contain all types

gauravsnj committed:
Update DjangoBasicTest.java

gauravsnj committed:
Update DjangoBasicTest.java

gauravsnj committed:
Merge branch 'postgresql-dialect' into django-extra-tests

gauravsnj committed:
Update DjangoBasicTest.java

gauravsnj committed:
Merge branch 'postgresql-dialect' into django-extra-tests

olavloite committed:
test: try with explicit time zone

olavloite commented on src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoBasicTest.java:
```diff
@@ -230,4 +262,92 @@ public void testInsertAllTypesAllNull() throws IOException, InterruptedException
+    String actualOutput = executeBasicTests(pgServer.getLocalPort(), host, options);
+    assertEquals(expectedOutput, actualOutput);
+
+    assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
```
nit: could we also have a verification here that the statement is executed using a single-use read-only transaction? So something like `assertTrue(request.getTransaction().hasSingleUse());`

gauravsnj replied:
done

olavloite commented on src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoBasicTest.java:
```diff
@@ -230,4 +262,92 @@ public void testInsertAllTypesAllNull() throws IOException, InterruptedException
     assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
     assertEquals(sqlInsert, mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0).getSql());
```
nit: could we add a verification that the statement is executed in auto-commit mode. That is:
1. The execute request includes a BeginTransaction option with a read/write-option=true.
2. There is a commit request on the server.

The reason for the above is that Django is a very big exception to most ORMs. Most ORMs will not use auto-commit by default, but will use read/write transactions that must be committed explicitly by default. We should verify this behavior in order to ensure that our implementation is correct, and to be able to detect potential regressions in the future if we or the Django maintainers make any changes.

gauravsnj replied:
done

olavloite reviewed (approved):
LGTM (once we have been able to fix the test failure)

gauravsnj committed:
Update DjangoBasicTest.java

gauravsnj committed:
Merge branch 'django-extra-tests' of https://github.com/GoogleCloudPlatform/pgadapter into django-extra-tests

gauravsnj committed:
Update DjangoBasicTest.java

gauravsnj committed:
test

gauravsnj committed:
Update AbstractMockServerTest.java

olavloite committed:
test: use timestamp with micros precision

olavloite committed:
test: use microsecond precision in timestamp

olavloite committed:
fix: revert to zulu time for mock Spanner

Cloud Spanner always returns timestamps in Zulu time. PGAdapter will
convert this into the PostgreSQL text format.

olavloite committed:
test: use same timezone setting for all tests

olavloite committed:
Merge branch 'postgresql-dialect' into django-extra-tests
